src/lib/y2configuration_management/salt/form.rb

```diff
@@ -0,0 +1,226 @@
+# more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, contact SUSE LLC.
+#
+# To contact SUSE LLC about this file by physical or electronic mail, you may
+# find current contact information at www.suse.com.
+
+require "yaml"
+
+module Y2ConfigurationManagement
+  module Salt
+    # Class that represents a form for Salt Formulas
+    class Form
+      # @return [Array<FormElement>]
+      attr_reader :elements
+      # @return [Hash]
+      attr_reader :spec
+
+      # Constructor
+      #
+      # @param spec [Hash]
+      def initialize(spec)
+        @elements = []
+        @spec = spec
+        build_elements
+      end
+
+      # Creates a new Form object reading the definition from a YAML file
+      #
+      # param path [String] file path to read the form YAML definition
+      def self.from_file(path)
+        definition = YAML.load(File.read(path))
+        new(definition)
+      end
+
+    private
+
+      # Return specification form elements
+      #
+      # @return [Hash]
+      def spec_elements
+        spec.select { |k, v| !k.start_with?("$") }
+      end
+
+      # Return the form attributes from the specification
+      #
+      # @return [Hash]
+      def attributes
+        spec.select { |k, v| k.start_with?("$") }
+      end
+
+      # It builds the form elements from the specification
+      def build_elements
+        spec_elements.each { |n, h| elements << FactoryFormElement.build(n, h, parent: self) }
+      end
+    end
+
+    # It builds new Form Elements depending on its specification type
+    class FactoryFormElement
+      # Builds a new FormElement object based on the element specification and
+      # maintaining a reference to its parent
+      #
+      # @param name [String]
+      # @param spec [Hash]
+      # @param parent [Form, FormElement]
+      def self.build(name, spec, parent: )
+        class_for(spec["$type"]).new(name, spec, parent: parent)
+      end
+
+      # @param type [String]
+      # @return [FormElement]
+      def self.class_for(type)
+        case type
+        when "namespace", "hidden-group", "group"
+          Container
+        when "edit-group"
+          Collection
+        else
+          FormInput
+        end
+      end
+    end
+
+    # Three different kind of elements:
+    #
+    # scalar values, groups and collections
+    class FormElement
+      # @return [Form, FormElement]
+      attr_reader :parent
+      # @return  [String]
+      attr_reader :name
+      # @return [String]
+      attr_reader :help
+      # @return scope [String] specify the level in which the value can be
+      #   edited. Possible values are: system, group and readonly
+      attr_reader :scope
+      # @return optional [Boolean]
+      attr_reader :optional
+
+      # Constructor
+      #
+      # @param name [String]
+      # @param spec [Hash] form element specification
+      def initialize(name, spec, parent: nil)
```

Comment:
IMHO leaving out the parent is an exception, not the usual case, so it should not have a default (but rather the root element should explicitly say `parent:nil`.
```suggestion
      def initialize(name, spec, parent:)
```